File: python/src/sec_edgar_toolkit/types/parsing.py
"""Type definitions for parsing infrastructure."""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional, TypedDict, Union


class DocumentSection(TypedDict):
    """Document section information."""

    name: str
    content: str
    start_index: int
    end_index: int
    type: str  # 'text', 'table', 'xbrl', 'exhibit'


class TableMetadata(TypedDict):
    """Table metadata information."""

    location: str
    format: str  # 'html', 'text', 'xbrl'


class Table(TypedDict):
    """Parsed table information."""

    title: str
    headers: List[str]
    rows: List[List[str]]
    metadata: TableMetadata


class Exhibit(TypedDict):
    """Exhibit information."""

    number: str
    title: str
    description: str
    filename: str
    type: str
    content: Optional[str]


class DocumentMetadata(TypedDict):
    """Document metadata information."""

    document_count: int
    has_xbrl: bool
    has_html: bool
    file_size: int
    processing_time: float
    parsing_errors: List[str]


class ParsedDocument(TypedDict):
    """Complete parsed document data."""

    form_type: str
    filing_date: datetime
    cik: str
    company_name: str
    accession_number: str
    raw_content: str
    sections: List[DocumentSection]
    tables: List[Table]
    exhibits: List[Exhibit]
    metadata: DocumentMetadata


class Metric(TypedDict):
    """Extracted metric information."""

    name: str
    value: Union[float, str]
    units: str
    confidence: float
    source: str


class Statement(TypedDict):
    """Extracted statement information."""

    type: str  # 'forward_looking', 'risk_factor', 'material_change'
    content: str
    confidence: float
    location: str


class ValidationError(TypedDict):
    """Validation error information."""

    type: str
    message: str
    location: str
    severity: str  # 'error', 'warning', 'info'


class ValidationWarning(TypedDict):
    """Validation warning information."""

    type: str
    message: str
    location: str


class ValidationResult(TypedDict):
    """Validation result information."""

    is_valid: bool
    errors: List[ValidationError]
    warnings: List[ValidationWarning]

File: typescript/src/types/parsing.ts
/**
 * Type definitions for parsing infrastructure
 */

export interface ParsedDocument {
  formType: string;
  filingDate: Date;
  cik: string;
  companyName: string;
  accessionNumber: string;
  rawContent: string;
  sections: DocumentSection[];
  tables: Table[];
  exhibits: Exhibit[];
  metadata: DocumentMetadata;
}

export interface DocumentSection {
  name: string;
  content: string;
  startIndex: number;
  endIndex: number;
  type: 'text' | 'table' | 'xbrl' | 'exhibit';
}

export interface Table {
  title: string;
  headers: string[];
  rows: string[][];
  metadata: {
    location: string;
    format: 'html' | 'text' | 'xbrl';
  };
}

export interface Exhibit {
  number: string;
  title: string;
  description: string;
  filename: string;
  type: string;
  content?: string;
}

export interface DocumentMetadata {
  documentCount: number;
  hasXBRL: boolean;
  hasHTML: boolean;
  fileSize: number;
  processingTime: number;
  parsingErrors: string[];
}

export interface Metric {
  name: string;
  value: number | string;
  units: string;
  confidence: number;
  source: string;
}

export interface Statement {
  type: 'forward_looking' | 'risk_factor' | 'material_change';
  content: string;
  confidence: number;
  location: string;
}

export interface ValidationResult {
  isValid: boolean;
  errors: ParsingValidationError[];
  warnings: ValidationWarning[];
}

export interface ParsingValidationError {
  type: string;
  message: string;
  location: string;
  severity: 'error' | 'warning' | 'info';
}

export interface ValidationWarning {
  type: string;
  message: string;
  location: string;
}